### ISSUE TITLE
Inside `impl`s, complete unimplemented functions

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#1046]
See https://github.com/rust-analyzer/rust-analyzer/issues/1022#issuecomment-475916555

The idea is that for

```rust
trait T {
   fn foo();
}

impl T for () {
   |
}
```

We should get completions for `foo`.  Specifically, the completion should trigger on `fn` keyword and on bare identifier (so that users who know about this feature can just type the identifier)

The code, responsible for completions, lives here:

https://github.com/rust-analyzer/rust-analyzer/blob/958df1276e3d97b189dcd9de974564011c70ef62/crates/ra_ide_api/src/completion.rs

To implement this feature, one has to add more info to `CompletionContext` and to create a separate completion module. Take a look at `complete_fn_param` which does somewhat similar magical completions. 

This should also work for `type`s and `const`s

**Comment by flodiebold** [comment]
Also, the completion should complete to the whole fn (/type/const) declaration, e.g. `fn foo() {|}` :) I assume that's implied, but wanted to make sure ;)

**Comment by detrumi** [comment]
I'll be working on this. Found the `IMPL_BLOCK` on `ctx.leaf.ancestors()`, so I know where to start :)

**Comment by hargoniX** [comment]
So the PR from detrumi has apparently been closed due to inactivity, does this mean that this issue is open to work on again?

**Comment by matklad** [comment]
Yes, go for it!

**Comment by hargoniX** [comment]
So having looked a bit into the code and the docs for this project, should I just take over the exact code from the original PR and continue working from there or should I write my own PR from the base up?

**Comment by kdelorey** [comment]
Since there hasn't been activity on this in a while, I'm going to attempt it if that's alright. I do have basic functionality working as a proof-of-concept to myself.

**Comment by matklad** [comment]
@kdelorey :+1:

**Comment by kdelorey** [comment]
Update, I'm still working on this slowly but surely on my free time. I've just been trying to get up to speed so I can accomplish this task in the effective way I can think of (which I think is relying on the hir as much as I can).

**Comment by kdelorey** [comment]
@matklad (or whomever), is there any recommendations in regards to the completion of an associated constant? 

I basically want the magic completion to write the following:
```rust
trait SomeTrait {
    const SOMETHING: u8;
}

impl SomeTrait for () {
    // text in between `[` and `]` is a magic completion:
    [const SOMETHING: u8 = ]
}
```
I'm mostly having trouble getting the type information in order to formulate the completion. It feels maybe a little dirty, but I could walk the syntax tree and copy everything up until the `;` (or `=` in cases where a default value is supplied)?

**Comment by matklad** [comment]
@kdelorey take a look how https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_assists/src/handlers/add_missing_impl_members.rs is implemented, I think that that intention and asssist might actually want to share code for "which functions/types/constatns are yet to be implemetned?"

**Comment by kdelorey** [comment]
> asssist might actually want to share code for "which functions/types/constatns are yet to be implemetned?"

I've also thought this, but wasn't sure where to put it. I could just export it from the `ra_assists` crate as `ra_ide` depends on it, I guess?

**Comment by matklad** [comment]
Right. Alternatively, we’ve introduced untils.rs module inside assists for
this stuff.

**Comment by kdelorey** [comment]
@matklad so based on what use said, I did [this](https://github.com/kdelorey/rust-analyzer/commit/d85abd77b98ff5925621c18f2ffe121640d17c80). It's a little bit different than the implementation of the assist's (`add_missing_impl_memebers`) missing members to resolve a bug I discovered. The bug is that that member names are currently evaluated together across the different allowed associated items, but the compiler technically allows different associated items to have the same name, regardless of it being idiomatic.

The following is allowed:

```
trait SomeTrait {
   type whatever;

   fn whatever();
}
```

I haven't updated the assist to use this method yet, but I wanted to run the location of the method and making the utils `pub` by you.

Thanks!

**Comment by matklad** [comment]
Yup, looks reasonable to me!

**Comment by kdelorey** [comment]
So I think it's basically all implemented. Just needs some cleanup before a PR.

![TraitCompletion](https://user-images.githubusercontent.com/2295721/74118816-ab4cf300-4b82-11ea-9b24-fb0b35021d62.gif)

I couldn't figure out how to change the cursor on a completion, but it would be nice if the function completion placed the cursor inbetween the curly braces of a `fn`.
